torchgen/dest/dispatchless.py

```diff
@@ -0,0 +1,265 @@
+        #   2. the CompositeExplicitAutograd kernel
+        #   3. the CompositeImplicitAutograd kernel
+        for k in (
+            self.dispatch_key,
+            DispatchKey.CompositeExplicitAutograd,
+            DispatchKey.CompositeImplicitAutograd,
+        ):
+            if has_registered_kernel(self.backend_indices[k], f, g):
+                return k
+
+        # bail if we have found no registered kernel.
+        return None
+
+    def _header_set(
+        self,
+        header_from_fn: Callable[[NativeFunction, Optional[NativeFunctionsGroup]], str],
+    ) -> List[str]:
+        return list(
+            {
+                header_from_fn(dep_f, dep_g)
+                for _, info in self.graph.items()
+                for dep_f, dep_g in info.dependencies
+            }
+        )
+
+    def aggregated_headers(self) -> List[str]:
+        def header_from(
+            dep_f: NativeFunction, dep_g: Optional[NativeFunctionsGroup]
+        ) -> str:
+            ns = self._dispatch_namespace(dep_f, dep_g)
+            prefix = f"{ns}" if ns is not None else ""
+            return f"#include <ATen/{prefix}Functions.h>"
+
+        return self._header_set(header_from)
+
+    def operator_headers(self) -> List[str]:
+        def header_from(
+            dep_f: NativeFunction, dep_g: Optional[NativeFunctionsGroup]
+        ) -> str:
+            ns = self._dispatch_namespace(dep_f, dep_g)
+            suffix = f"_{ns.lower()}_dispatch" if ns is not None else ""
+            return f"#include <ATen/ops/{dep_f.root_name}{suffix}.h>"
+
+        return self._header_set(header_from)
+
+    def headers(self) -> List[str]:
+        native_functions = [info.f for _, info in self.graph.items()]
+        return sorted(
+            {
+                f"#include <ATen/native/composite/{f.root_name}.h>"
+                for f in native_functions
+            }
+        )
+
+    @method_with_native_function
+    def definition(self, f: NativeFunction) -> str:
+        if f.func.name not in self.graph:
+            return ""
+
+        # define the native function as a wrapper static method, redirecting the call to
+        # the appropriate function (e.g. on CPU, on CUDA, or as a CompositeExplicitAutograd).
+        def dispatchless_function_defn(
+            dep_f: NativeFunction, dep_g: Optional[NativeFunctionsGroup]
+        ) -> str:
+            with native_function_manager(dep_f):
+                ns = self._dispatch_namespace(dep_f, dep_g)
+                if ns is None:
+                    # if we haven't found a namespace, we defer it to the dispatcher.
+                    # however, there's a possibility that such an operation is only available
+                    # as a tensor method.
+                    if Variant.function in dep_f.variants:
+                        prefix = "at::"
+                        method = False
+                    else:
+                        assert Variant.method in dep_f.variants
```

Comment:
also agreed that the special casing for method handling seems kinda necessary. At first I was gonna say that you could try having it call `at::_ops::foo::call` instead of `tensor.foo()`, but then you'd have an annoying API inconsistency (the _ops code all follows the dispatcher API)